pallets/pool-registry/src/lib.rs

```diff
@@ -0,0 +1,375 @@
+		/// added with the Permissions pallet.
+		///
+		/// Returns an error if the requested pool ID is already in
+		/// use, or if the tranche configuration cannot be used.
+		#[pallet::weight(T::WeightInfo::create(tranche_inputs.len().try_into().unwrap_or(u32::MAX)))]
+		#[transactional]
+		pub fn create(
+			origin: OriginFor<T>,
+			admin: T::AccountId,
+			pool_id: T::PoolId,
+			tranche_inputs: Vec<
+				TrancheInput<T::Rate, T::MaxTokenNameLength, T::MaxTokenSymbolLength>,
+			>,
+			currency: T::CurrencyId,
+			max_reserve: T::Balance,
+			metadata: Option<Vec<u8>>,
+		) -> DispatchResult {
+			T::PoolCreateOrigin::ensure_origin(origin.clone())?;
+
+			// First we take a deposit.
+			// If we are coming from a signed origin, we take
+			// the deposit from them
+			// If we are coming from some internal origin
+			// (Democracy, Council, etc.) we assume that the
+			// parameters are vetted somehow and rely on the
+			// admin as our depositor.
+			let depositor = ensure_signed(origin).unwrap_or(admin.clone());
+
+			T::ModifyPool::create(
+				admin.clone(),
+				depositor,
+				pool_id,
+				tranche_inputs,
+				currency,
+				max_reserve,
+				metadata,
+			)
+			.map(|_| Self::deposit_event(Event::Created { pool_id, admin }))
+		}
+
+		/// Update per-pool configuration settings.
+		///
+		/// This updates the tranches of the pool,
+		/// sets the minimum epoch length, and maximum NAV age.
+		///
+		/// If no delay is required for updates and redemptions
+		/// don't have to be fulfilled, then this is executed
+		/// immediately. Otherwise, the update is scheduled
+		/// to be executed later.
+		///
+		/// The caller must have the `PoolAdmin` role in order to
+		/// invoke this extrinsic.
+		#[pallet::weight(T::WeightInfo::update_no_execution(T::MaxTranches::get())
+		.max(T::WeightInfo::update_and_execute(T::MaxTranches::get())))]
+		pub fn update(
+			origin: OriginFor<T>,
+			pool_id: T::PoolId,
+			changes: PoolChanges<
+				T::Rate,
+				T::MaxTokenNameLength,
+				T::MaxTokenSymbolLength,
+				T::MaxTranches,
+			>,
+		) -> DispatchResultWithPostInfo {
+			// Make sure the following are true for a valid Pool Update
+			// 1. Make sure Origin is signed
+			// 2. Ensure the signed origin is PoolAdmin
+			// 3. Either no changes in tranches/metadata or changes in both
+			// 4. Ensure not in Submission Epoch Time
+			// 5. If no changes, do no_op
+			// 6. Minimum epoch time has to be between min and max epoch time
+			// 7. MaxNavAge has to be under the upper bound NavAge
+			// 8. IsValidTrancheChange?
+
+			let who = ensure_signed(origin)?;
+			ensure!(
+				T::Permission::has(
+					PermissionScope::Pool(pool_id),
+					who,
+					Role::PoolRole(PoolRole::PoolAdmin)
+				),
+				BadOrigin
+			);
+
+			match T::ModifyPool::update(pool_id, changes) {
+				Ok((state, dispatch_info)) => {
+					Self::deposit_event(Event::UpdateRegisted { pool_id });
+
+					match state {
+						UpdateState::NotExecution => {
```

Comment:
I'm not sure if we need to emit an event in case of a no-op. I think the core goal of events is for clients to easily figure out what changed; if nothing changed, it doesn't seem like this requires an event. Wdyt?